### DISCUSSION TITLE
How to load engines dynamically from config?

### DISCUSSION THREAD

**Discussion by Sourabh-ALTEN** [openssl/openssl#24586] [Q&A]
Hello All,

This is rather a question than issue.

Currently we are loading a pkcs11 engine dynamically from our source code by calling `e = ENGINE_by_id("dynamic");` and `ENGINE_ctrl_cmd_string`. Now we want to make use of `openssl.cnf` file to do this. I wanted to know whether only providing `dynamic_path` and `MODULE_PATH` in `openssl.cnf` is sufficient to tell openssl to use this pkcs11 engine? Do I still need to add ENGINE section in our source code?

And also, if we specify `PIN` in openssl.cnf file then how can we access it within our code? For example, what if we want to use the variable defined in config file? Any example or documentation for this would be helpful.

```
[engine_section] 
pkcs11 = pkcs11_section

[pkcs11_section]
engine_id = pkcs11 
dynamic_path = /usr/lib/engines-1.1/pkcs11.so
MODULE_PATH = /usr/lib/libtpm2_pkcs11.so
PIN = 12xx
default_algorithms = ALL
init = 0
```

Thanks in advance.

P.S: Please let me know if further info is needed.

**Comment by beldmit** [comment]
Here is the example for GOST engine. You need to have a `openssl_conf` section

```
openssl_conf = openssl_def
[openssl_def]
engines = engine_section

[engine_section]
gost = gost_section

[gost_section]
engine_id = gost
dynamic_path = /path/to/gost.so
default_algorithms = ALL
```

**Comment by Sourabh-ALTEN** [comment]
Hello @beldmit thanks for quick response.

Yes, we have already included this `openssl_conf` section. But I wanted to know how this is parsed by openssl. And also do we need to still initialize newly loaded engine using `ENGINE_init()`?

**Comment by beldmit** [comment]
No, you don't need to call `ENGINE_init()` directly
You can try the `openssl engine` command for some diagnostics purposes

**Comment by Sourabh-ALTEN** [comment]
Okay, If I understand it correctly then we don't have to use any `ENGINE_ctrl_cmd_string` command also if everything is mentioned in openssl_conf section and this is taken care by openssl. But, if we specify PIN in `openssl.cnf` file then how can we access it within our code? For example, what if we want to check the PIN value which is defined in config file?

**Comment by beldmit** [comment]
`NCONF_get_string` function may help for getting value from the config, but I didn't check it here and now.
If the parameter is accessable via ENGINE API, it may also be an option, though I'm not sure it's the case.
